### PR TITLE
gh-482 Add OAuth2 Support

### PR DESCRIPTION
**Pull request by ghillert**
* Add documentation
  - Add documentation on how to setup SSL
  - Add documentation on how to setup authentication using OAuth2
* Add shell support
* Add tests for shell support
* Add `org.apache.httpcomponents:httpclient` to `spring-cloud-dataflow-rest-client` in order to enable basic authentication support
* Add the following dependencies to `spring-cloud-dataflow-server-core`:
  - spring-boot-starter-security
  - spring-security-oauth2
  - spring-cloud-sso-connector
* Add `/security/info` controller

Resolves spring-cloud/spring-cloud-dataflow#482
Resolves spring-cloud/spring-cloud-dataflow-ui#49

### CODE REVIEW

**Comment by ghillert** [comment]
*Question*:

Right now I keep all security dependencies in one spot `spring-cloud-dataflow-server-core/pom.xml`. Should the following `spring-cloud-sso-connector`:

```
		<dependency>
			<groupId>io.pivotal.spring.cloud</groupId>
			<artifactId>spring-cloud-sso-connector</artifactId>
			<version>1.1.0.RELEASE</version>
		</dependency>
```

Be better moved to `https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry`?

**Comment by ilayaperumalg** [comment]
@ghillert I think it would be good to move the `spring-cloud-sso-connector` dependency to `spring-cloud-dataflow-server-cloudfoundry` as it is not used by any other server implementations.

**Comment by ghillert** [comment]
@ilayaperumalg Will do. That will also preserve Java `1.7` compatibility. Otherwise `spring-cloud-sso-connector` requires `1.8`.

**Comment by ghillert** [comment]
Created PR for `spring-cloud-dataflow-server-cloudfoundry` adding the `spring-cloud-sso-connector` dependency there:

https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry/pull/106

**Comment by ghillert** [comment]
Removed `spring-cloud-sso-connector` dependency

**Comment by ghillert** [comment]
@ilayaperumalg Added more documentation for setting up SSL including step-by-step instructions for using self-signed certs.

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc [hunk 1]
```diff
@@ -66,37 +67,214 @@ export MAVEN_PROXY_PORT=8090
...
+----
+security:
+  basic:
+    enabled: true                                                     # <1>
+    realm: Spring Cloud Data Flow                                     # <2>
+  oauth2:                                                             # <3>
+    client:
+      client-id: myclient
+      client-secret: mysecret
+      access-token-uri: http://127.0.0.1:9999/oauth/token
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc]: @ghillert How do I set the oauth server up locally?

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc [hunk 2]
```diff
@@ -66,37 +67,214 @@ export MAVEN_PROXY_PORT=8090
...
+      user-authorization-uri: http://127.0.0.1:9999/oauth/authorize
+    resource:
+      user-info-uri: http://127.0.0.1:9999/me
+----
+
+<1> Must be set to `true` for security to be enabled.
+<2> The realm for Basic authentication
+<3> OAuth Configuration Section
+
+NOTE: As of version 1.0 Spring Cloud Dataflow does not provide finer-grained authorization. Thus, once you are logged in, you have full access to all functionality.
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc]: When using `OAuth`, how can the dataflow shell connect or how can we specify the authentication parameters?

File: spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc (hunk 1 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/getting-started.adoc]: Got it from here: https://github.com/ghillert/oauth-test-server/tree/master thanks @ghillert

**Comment by ilayaperumalg** [comment]
LGTM, also verified Dataflow `local` server/shell with the test `OAuth` server running locally. merging.

**Comment by ilayaperumalg** [comment]
Squashed and merged as cc61242dce57d41dc064f630034a6782cf82d158

**Pull request closed without merging**